tests/exact/test_branch_and_bound.py

```diff
@@ -0,0 +1,189 @@
+
+    assert response_node.level == 0
+    assert response_node.index == 0
+    assert response_node.cost == 25
+    assert np.all(response_node.reduced_matrix == reduced_cost_matrix)
+    assert response_node.path == [0]
+
+
+@pytest.mark.parametrize(
+    "node_index, node_expected_cost", [(1, 35), (2, 53), (3, 25), (4, 31)]
+)
+def test_create_child_node(cost_matrix, node_index, node_expected_cost):
+    """
+    Checks if the `Node.from_parent_node` method correctly creates a
+    new child Node based on a parent Node and a specified index. It
+    verifies if the attributes of the created child Node match the expected
+    values.
+
+    The test creates a root Node using the provided cost matrix, and then
+    generates a new child Node based on the root Node and the specified
+    node_index. The child Node's attributes are compared with the
+    expected values.
+    """
+    root = Node.root_from_cost_matrix(cost_matrix=cost_matrix)
+    response_node = Node.from_parent_node(
+        parent_node=root, node_index=node_index
+    )
+
+    assert response_node.level == 1
+    assert response_node.index == node_index
+    assert response_node.cost == node_expected_cost
+    assert response_node.path == [root.index, response_node.index]
+
+
+def test_nodes_priority_queue(cost_matrix):
+    """
+    Checks the functionality of the NodePriorityQueue by creating a priority
+    queue and adding child nodes to it based on the root node. It then
+    verifies if the queue is not empty and if the first node popped from the
+    queue has the expected cost value.
+
+    The test ensures that the priority queue correctly maintains the order
+    of Node objects based on their priority (cost) and pops the node with
+    the highest priority (lowest cost) first.
+    """
+    node_indices = range(len(cost_matrix))
+    pq = NodePriorityQueue()
+    root = Node.root_from_cost_matrix(cost_matrix=cost_matrix)
+
+    for node_index in node_indices:
+        is_live_node = root.reduced_matrix[root.index][node_index] != INF
+        if is_live_node:
+            live_node = Node.from_parent_node(
+                parent_node=root, node_index=node_index
+            )
+            pq.push(live_node)
+
+    assert not pq.empty
+    assert pq.pop().cost == 25
+
+
+@pytest.mark.parametrize(
+    "distance_matrix", [distance_matrix1, distance_matrix2, distance_matrix3]
+)
+def test_solution_has_all_nodes(distance_matrix):
+    """Check if the solution has all input nodes in any order"""
+
+    permutation, _ = solve_tsp_branch_and_bound(distance_matrix)
+
+    num_nodes = distance_matrix.shape[0]
+    assert len(permutation) == num_nodes
+    assert set(permutation) == set(range(num_nodes))
+
+
+@pytest.mark.parametrize(
+    "distance_matrix, expected_distance",
+    [
+        (distance_matrix1, optimal_distance1),
+        (distance_matrix2, optimal_distance2),
+        (distance_matrix3, optimal_distance3),
+    ],
+)
```

Comment:
@fillipe-gsm For test case 2 the algorithm finds the permutation [0, 1, 4, 3, 2] also with an optimal cost of 21, but different from the expected permutation that it fixed at [0, 1, 2, 4, 3]. That's why I didn't make claims about equality of permutations.

Reply:
@fillipe-gsm but `test_solution_has_all_nodes` guarantees that solutions found for these same entries are valid permutations!